addon/addon/models/animation-participant.ts

```diff
@@ -32,39 +32,129 @@ export class AnimationParticipantManager {
+          participant.uiState.detached.animation?.playState === 'paused')
+      ) {
+        // TODO: when cloning is implemented, add the DOMRef here
+        unanimatedDetachedDOMRefs.add(participant.uiState.detached.DOMRef);
+      }
+    }
 
+    let deleted: DOMRefNode[] = [];
+    let nodesToGraft = new Map<DOMRefNode, DOMRefNode>();
+    function recurse<T>(
+      node: DOMRefNode,
+      callback: (node: DOMRefNode, scope: T) => { nextScope: T; stop: boolean },
+      scope: T
+    ) {
+      let { nextScope, stop } = callback(node, scope);
+      if (stop) return;
+      for (let child of node.children) {
+        recurse(child, callback, nextScope);
+      }
+    }
+    for (let node of displacedDOMRefs) {
+      recurse<undefined>(
+        node,
+        (node, _) => {
+          if (node.parent) node.delete();
+          else this.DOMRefs = this.DOMRefs.filter((v) => v !== node);
+          deleted.push(node);
+          return {
+            nextScope: undefined,
+            stop: false,
+          };
+        },
+        undefined
+      );
+    }
+    for (let DOMRef of this.DOMRefs) {
+      recurse<{
+        lastLiveAncestor: DOMRefNode | undefined;
+        isOnDetachedBranch: boolean;
+        stop: boolean;
+      }>(
+        DOMRef,
+        (node, scope) => {
+          if (unanimatedDetachedDOMRefs.has(node)) {
+            if (node.parent) node.delete();
+            else this.DOMRefs = this.DOMRefs.filter((v) => v !== node);
+            deleted.push(node);
+            return {
+              nextScope: {
+                ...scope,
+                isOnDetachedBranch: true,
+              },
+              stop: false,
+            };
+          } else if (scope.isOnDetachedBranch) {
+            if (scope.lastLiveAncestor) {
+              nodesToGraft.set(node, scope.lastLiveAncestor);
+              return {
+                nextScope: {
+                  ...scope,
+                },
+                stop: true, // no need to iterate into the children of this node, because it is currently being used (it has animations on it)
+              };
+            } else {
+              if (node.parent) node.delete();
+              else this.DOMRefs = this.DOMRefs.filter((v) => v !== node);
+              deleted.push(node);
+              return {
+                nextScope: {
+                  ...scope,
+                },
+                stop: false,
+              };
+            }
+          } else {
+            // We're still on a live branch, so return this node as the lastLiveAncestor
+            // It will become the graft target later
+            return {
+              nextScope: {
+                ...scope,
+                lastLiveAncestor: node,
+              },
+              stop: false,
+            };
+          }
+        },
+        {
+          isOnDetachedBranch: false,
+          lastLiveAncestor: undefined,
+          stop: false,
+        }
+      );
```

Comment:
Iterate over all DOMRefNodes from top to bottom, removing anything that was marked as detached and not animating, or ready for disposal.

While iterating over descendants of nodes that are marked as detached and not animating, if we hit something that's animating, we need to keep it around. We want to graft it, as a child of the last (lowest) live ancestor. We'll do the grafting later, we mark it for now.